### ISSUE TITLE
Proposal for a method that creates a branch of nodes if it doesn't exist yet

### ISSUE THREAD

**Issue by funkyfuture** [funkyfuture/delb#18]
particularly for metadata, but also witnessed for translations, there's a rather bloated pattern to ensure that a full branch of nodes exists:

```python
root = Document("<root/>").root
if not root.xpath("./foo"):
    root.append_child(tag("foo"))
if not root.xpath("./foo/bar"):
    root.xpath("./foo").first.append_child(tag("bar"))
bar = root.xpath("./foo/bar").first
```

the deeper this goes, the more tedious it gets.

the idea is to enable something like this:

```python
root = Document("<root/>").root
bar = root.get_or_create_by_xpath("./foo/bar")
```

the intrinsic assumption would be that only one node with a given name exists per location step.

shall attribute-test with distinct values also be considered? that'd be neat, e.g. `./titleStmt/title[@type="main"]` and `./titleStmt/title[@type="sub"]`.

the good thing is that there's already an internally used data model for XPath expressions that can be extended.

a variant for css selector based expressions would just use the translations of such like the query methods.

for a first release of this feature, it will be marked as experimental.

**Comment by JKatzwinkel** [comment]
What would happen if there exists one element under xpath `./titleStmt/title`, and I call `get_or_create('./titleStmt/title[@type="main"]')`? Do I end up with one or two elements matching `./titleStmt/title`?

**Comment by funkyfuture** [comment]
42!

no, two. if the existing element doesn't bear said attribute *and* value.

**Comment by funkyfuture** [comment]
implemented w/ 432238864f92943618ad27be53739559580866f0
